packages/pubsub/src/queue.js

```diff
@@ -0,0 +1,200 @@
+/*!
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module pubsub/queue
+ */
+
+'use strict';
+
+var common = require('@google-cloud/common');
+var prop = require('propprop');
+
+/**
+ * Semi-generic Queue object used to queue up multiple requests while limiting
+ * the number of concurrent requests.
+ *
+ * @constructor
+ * @alias module:pubsub/queue
+ *
+ * @param {object} options - Queue configuration settings.
+ * @param {number} options.maxBytes - Max size of data to be sent per request.
+ * @param {number} options.maxMessages - Max number of messages to be sent per
+ *     request.
+ * @param {number} options.maxMilliseconds - Max time to wait before sending
+ *     data.
+ * @param {number} options.maxRequests - Max number of concurrent requests
+ *     allowed.
+ * @param {function(messages, callback)} options.sendHandler - Function to be
+ *     called with message queue.
+ *
+ * @example
+ * var queue = new Queue({
+ *   maxBytes: Math.pow(1024, 2) * 5,
+ *   maxMessages: 1000,
+ *   maxMilliseconds: 1000,
+ *   maxRequests: 5,
+ *   send: function(messages, done) {
+ *     var reqOpts = {
+ *       topic: self.topic.name,
+ *       messages = messages.map(function(message) {
+ *         return {
+ *           data: message.data,
+ *           attributes: message.attrs
+ *         };
+ *       })
+ *     };
+ *
+ *     return self.api.Publisher.publish(reqOpts, function(err, resp) {
+ *       if (err) {
+ *         done(err);
+ *         return;
+ *       }
+ *
+ *       done(null, {
+ *         responses: resp.messageIds
+ *       })
+ *     });
+ *   }
+ * });
+ */
+function Queue(options) {
+  this.maxBytes = options.maxBytes || Math.pow(1024, 2) * 5;
```

Comment:
Actually I would prefer it be a little lower, because we do not really have a sufficiently precise understanding of how many bytes are in a message.

I would do `Math.min(options.MaxBytes || Math.pow(1024, 2) * 5, Math.pow(1024, 2) * 9)`.